src/account/UpgradeableModularAccount.sol

```diff
@@ -692,14 +701,58 @@ contract UpgradeableModularAccount is
         return getAccountStorage().selectorData[selector].allowGlobalValidation;
     }
 
-    function _checkPermittedCallerIfNotFromEP() internal view {
+    /**
+     * Order of operations:
+     *      1. Check if the sender is the entry point, the account itself, or the selector called is public.
+     *          - Yes: Return an empty array, there are no post-permissionHooks.
+     *          - No: Continue
+     *      2. Check if the called selector (msg.sig) is included in the set of selectors the msg.sender can
+     *         directly call.
+     *          - Yes: Continue
+     *          - No: Revert, the caller is not allowed to call this selector
+     *      3. If there are runtime validation hooks associated with this caller-sig combination, run them.
+     *      4. Run the pre-permissionHooks associated with this caller-sig combination, and return the
+     *         post-permissionHooks to run later.
+     */
+    function _checkPermittedCallerAndAssociatedHooks()
+        internal
+        returns (PostExecToRun[] memory, PostExecToRun[] memory)
+    {
         AccountStorage storage _storage = getAccountStorage();
 
         if (
-            msg.sender != address(_ENTRY_POINT) && msg.sender != address(this)
-                && !_storage.selectorData[msg.sig].isPublic
+            msg.sender == address(_ENTRY_POINT) || msg.sender == address(this)
+                || _storage.selectorData[msg.sig].isPublic
         ) {
+            return (new PostExecToRun[](0), new PostExecToRun[](0));
+        }
+
+        FunctionReference directCallValidationKey =
+            FunctionReferenceLib.pack(msg.sender, _SELF_PERMIT_VALIDATION_FUNCTIONID);
+
+        if (!_storage.validationData[directCallValidationKey].selectors.contains(toSetValue(msg.sig))) {
             revert ExecFromPluginNotPermitted(msg.sender, msg.sig);
         }
```

Comment:
Following up on a conversation from here: https://github.com/erc6900/reference-implementation/pull/78#issuecomment-2187328143

There's a risk if the direct call path calls `execute`/`executeBatch` with the target of self again, because it would allow the plugin performing the direct call to execute selectors that were not granted to it.

I think we should update this call to be
```solidity
if (!_checkIfValidationAppliesCallData(msg.data, directCallValidationKey, false)) {
```

This would allow for the self-call batch operation via direct calls, but no privilege escalation.

Alternatively, we could fully ban any self-directed calls from a plugin direct calling `execute`/`executeBatch`. (wow that's a moutful). I don't think the use case of batching via `executeBatch` will be that common, but it would allow us to consolidate the code for checking validation applicability.

Reply:
Ah a small issue with this-- I agree, but in the `_checkIfValidationAppliesCallData()` function, in the `executeBatch()` scenario, we check that the validation applies to the inner selector-- which means we'd have to add the selectors to this permit-all validation.

This made me realize that there's a limitation if we want the plugin to be able to call selector A with a set of associated hooks, and selector B with another set of associated hooks, right?

We can't re-install the validation because the key has to be the same, and it's used a storage key. Is this a real issue or am I tripping?

Either way, great catch! Implementing this change now.

Reply:
Addressed here: [a978cd7](https://github.com/erc6900/reference-implementation/pull/90/commits/a978cd7d307f27652e31ee7d2e1d508109dd3366)

Lmk if it looks good!

Reply:
> This made me realize that there's a limitation if we want the plugin to be able to call selector A with a set of associated hooks, and selector B with another set of associated hooks, right?
We can't re-install the validation because the key has to be the same, and it's used a storage key. Is this a real issue or am I tripping?

Yes, this is the case. Every address can only have 1 config (domain [selectors, global, isSigValidation] + hooks [pre-validation, permission]) for the direct call path.

However, I think it does kinda make sense - in the direct call path, we don't have any information to select the "config" from other than the caller address. We can't reliably encode an ID anywhere, if we want to call the native functions as normal and support arbitrary ABI functions in the fallback. So there's only one ID we can use, which is designated as the magic value in the code & in storage.

Reply:
> which means we'd have to add the selectors to this permit-all validation.

I think this is OK, right? If the plugin really depends on this, it would need to request access to N selectors, and would then be able to batch call them directly?

Reply:
> Yes, this is the case. Every address can only have 1 config (domain [selectors, global, isSigValidation] + hooks [pre-validation, permission]) for the direct call path.

Makes sense to me, thanks! Let's leave it as is, I don't see any reasonable alternative and I think this will cover the vast majority of usecases!